manticore/platforms/linux.py

```diff
@@ -2115,21 +2115,36 @@ def sys_accept4(self, sockfd, addr, addrlen, flags):
         fd = self._open(sock)
         return fd
 
-    def sys_recv(self, sockfd, buf, count, flags):
-        try:
-            sock = self.files[sockfd]
-        except IndexError:
-            return -errno.EINVAL
+    def sys_recv(self, sockfd, buf, count, flags, trace_str="_recv"):
+        data: bytes = bytes()
+        if count != 0:
+            if buf not in self.current.memory or (buf + count) not in self.current.memory:
+                logger.info("buf or (buf + count) points to invalid address. Returning EFAULT")
+                return -errno.EFAULT
 
-        if not isinstance(sock, Socket):
-            return -errno.ENOTSOCK
+            try:
+                sock = self._get_fd(sockfd)
+            except FdError:
+                return -errno.EINVAL
```

Comment:
I notice that we only get to this case if `count != 0`. Does that mirror the real-life behavior of the `recv` syscall? Or should it be as it was in the previous version, where this call might return an error even if `count == 0`?

Reply:
Good catch.

According to the Linux kernel source, a check is done first on whether the whole buffer points to a valid place in memory, and then checks if the `fd` is valid. https://elixir.bootlin.com/linux/v5.2.13/source/net/socket.c#L2001